### ISSUE TITLE
Docker 1.12.4 creates race condition for GitLab-CI Runners

### ISSUE THREAD

**Issue by mrburrito** [docker/docker#29426]
**Description**

I'm using GitLab-CI with `docker-machine`-based runners to execute build jobs. After the update to 1.12.4 (which `docker-machine` applied automatically), all of my build jobs began to hang at various points until they timed out. This problem was traced back to the Docker update; runners work as expected in Docker 1.12.3.

GitLab is tracking the issue at https://gitlab.com/gitlab-org/gitlab-ci-multi-runner/issues/1962, and is working on a fix.

This [comment](https://gitlab.com/gitlab-org/gitlab-ci-multi-runner/issues/1962#note_20003535) indicates it's a race condition between `docker start` and `docker attach`.

There is also discussion of this issue in https://github.com/docker/machine/issues/1702

**Steps to reproduce the issue:**
1.  Standup GitLab with GitLab-CI
2. Register a GitLab-CI runner configured to use docker-machine
3. Trigger any CI job on the runner

**Describe the results you received:**

Build process would spin up the target container and begin executing build scripts but hang until GitLab's process timeout (1 hour) was exceeded.

**Describe the results you expected:**

Build process completes successfully inside container running on remote docker-machine host.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      1.12.4
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   1564f02
 Built:        Mon Dec 12 23:41:49 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.4
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   1564f02
 Built:        Mon Dec 12 23:41:49 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

This is from a similarly configured EC2 instance; not one of the failing runners. (Already configured a workaround)

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 1.12.4
Storage Driver: devicemapper
 Pool Name: docker-202:1-33587094-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 11.8 MB
 Data Space Total: 107.4 GB
 Data Space Available: 14.02 GB
 Metadata Space Used: 581.6 kB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.147 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.135-RHEL7 (2016-09-28)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-514.2.2.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 991.2 MiB
Name: dm-autoupgrade-test
ID: ISXJ:C2EQ:4LQ3:7CDQ:PKXG:VWSG:N3UL:K2BR:RREN:PSTE:FZ3L:YBPC
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
 provider=amazonec2
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

`docker-machine` executed by `gitlab-runner`, running in a container on an Amazon ECS cluster; uses `amazonec2` driver to spin up the machine(s) for each build task; base AMI is CentOS 7.3.1611.

**Comment by thaJeztah** [comment]
Thanks @mrburrito, much appreciated!

**Comment by tonistiigi** [comment]
This seems to be duplicate of #29421 . Fixed with #28682/#29424

The reason why the workaround works is that moves the sleep after container has started. The race happens when stdin from client is closed after the process has been already running.

**Comment by thaJeztah** [comment]
docker 1.12.5 has been released, which contains https://github.com/docker/docker/pull/29424 to resolve this issue; https://github.com/docker/docker/releases/tag/v1.12.5, so I'm closing this
